lib/Basics/process-utils.cpp

```diff
@@ -225,9 +225,115 @@ static bool CreatePipes(int* pipe_server_to_child, int* pipe_child_to_server) {
 /// @brief starts external process
 ////////////////////////////////////////////////////////////////////////////////
 
-static void StartExternalProcess(ExternalProcess* external, bool usePipes,
-                                 std::vector<std::string> const& additionalEnv,
-                                 std::string const& fileForStdErr) {
+#include <spawn.h>
+
+static void StartExternalProcessPosixSpawn(
+    ExternalProcess* external, bool usePipes,
+    std::vector<std::string> const& additionalEnv,
+    std::string const& fileForStdErr) {
+  int pipe_server_to_child[2];
+  int pipe_child_to_server[2];
+
+  if (usePipes) {
+    bool ok = CreatePipes(pipe_server_to_child, pipe_child_to_server);
+
+    if (!ok) {
+      external->_status = TRI_EXT_PIPE_FAILED;
+      return;
+    }
+  }
+
+  posix_spawn_file_actions_t file_actions;
+  posix_spawn_file_actions_init(&file_actions);
+  // file actions are performed in order they were added.
+
+  if (usePipes) {
+    posix_spawn_file_actions_adddup2(&file_actions, pipe_server_to_child[0], 0);
+    posix_spawn_file_actions_adddup2(&file_actions, pipe_child_to_server[1], 1);
+
+    posix_spawn_file_actions_addclose(&file_actions, pipe_server_to_child[0]);
+    posix_spawn_file_actions_addclose(&file_actions, pipe_server_to_child[1]);
+    posix_spawn_file_actions_addclose(&file_actions, pipe_child_to_server[0]);
+    posix_spawn_file_actions_addclose(&file_actions, pipe_child_to_server[1]);
+  } else {
+    posix_spawn_file_actions_addopen(&file_actions, 0, "/dev/null", O_RDONLY,
+                                     0);
+  }
+
+  if (!fileForStdErr.empty()) {
+    posix_spawn_file_actions_addopen(&file_actions, 2, fileForStdErr.c_str(),
+                                     O_CREAT | O_WRONLY | O_TRUNC, 0644);
+  }
+
+  posix_spawnattr_t spawn_attrs;
+  posix_spawnattr_init(&spawn_attrs);
+
+  std::vector<char*> envs;
+  for (char** e = environ; *e != nullptr; ++e) {
+    envs.push_back(*e);
+  }
+
+  envs.reserve(additionalEnv.size() + 1);
+  std::transform(
+      additionalEnv.begin(), additionalEnv.end(), std::back_inserter(envs),
+      [](auto& str) -> char* { return const_cast<char*>(str.data()); });
+  envs.emplace_back(nullptr);
+
+  int result = posix_spawnp(&external->_pid, external->_executable.c_str(),
+                            &file_actions, &spawn_attrs, external->_arguments,
+                            envs.data());
+
+  posix_spawnattr_destroy(&spawn_attrs);
+  posix_spawn_file_actions_destroy(&file_actions);
+
+  bool executableNotFound = false;
+
+  if (result != 0) {
+    if (errno == ENOENT) {
```

Comment:
I am not very happy about `errno` being used here for checking the error code returned by the failed `posix_spawnp` call. While checking `errno` is correct, I would prefer to do it directly after the call to `posix_spawnp`. Otherwise it is not guaranteed that the calls to `posix_spawnattr_destroy` or `posix_spawn_file_actions_destroy` do not clobber `errno` and overwrite it with something else. 
The same applies for the usage of `errno` later in the `strerror(errno)`. 
I suggest to simply assign the original `errno` value to a variable, and then use that variable later.